Add support for EDU-CIAA-NXP board

dc740 opened the pull request:
I'm finally wrapping up this port. This PR adds support for the board from tockloader.
Board description:
http://www.proyecto-ciaa.com.ar/index_comprar_en.html

By adding this I can now use

`tockloader  <command> --board edu-ciaa --openocd`

I wonder if I should add a way to be able to skip the --openocd flag. The board definition should be more than enough for this, but apparently it's required with the current implementation.

Known issues: I'm only being able to upload ONE single application. I think the `erase` command is actually erasing a bigger sector, so it effectively borks the second application install. I tried  removing it, and in some random cases I was able to upload 2 apps, but only the first one was running correctly anyway.

bradjc commented on tockloader/board_interface.py:
```diff
@@ -78,6 +78,15 @@ class BoardInterface:
 		            'arch': 'rv32imac',
 		            'page_size': 512,
 		            'openocd': 'sifive-hifive1.cfg'},
+		'edu-ciaa': {'description': 'Educational NXP board, from the CIAA proyect',
```
```suggestion
		'edu-ciaa': {'description': 'Educational NXP board, from the CIAA project',
```

dc740 replied:
thanks! I mixed it with the Spanish translation "proyecto"

bradjc commented:
Is it possible the flash page size is not actually 512 bytes? I've had issues with the nRF52 boards which have 4096 byte pages with apps that are smaller than 4096 bytes causing apps to mysteriously get erased. My fix was to ensure that apps on that chip are at least 4096 bytes long:

https://github.com/tock/tockloader/blob/7d0de0bd725a7794fc4ac0c57937f47e93791618/tockloader/tockloader.py#L61-L66

dc740 commented:
thanks for the suggestion. I will try it and let you know ASAP.

dc740 commented:
I'm going to need more time to debug it. It didn't work. I'll try to get to it by the weekend.

I'll leave the install log for the second app, just in case you see something extremely obvious. The first app is the blink example (I don't specify the tab, I let it download it from the web). Both apps get "uninstalled" after this. They are probably stepping on each other, but I'll take a look later with more time and patience.

```$ tockloader install ./build/buttons.tab --openocd --board edu-ciaa --debug
[INFO   ] Using settings from KNOWN_BOARDS["edu-ciaa"]
[STATUS ] Installing app on the board...
[DEBUG  ] Using read command: "dump_image {{binary}} {address:#x} {length};"
[DEBUG  ] Expanded read command: "dump_image {binary} 0x1a040000 200;"
[DEBUG  ] Running "openocd -c " source [find board/ftdi_lpc4337.cfg]; init; halt; dump_image /tmp/tmpv3zy0acc.bin 0x1a040000 200;  exit"".
[DEBUG  ] Open On-Chip Debugger 0.10.0+dev-00689-g6c2020eb (2019-02-10-17:40)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
adapter speed: 2000 kHz
none separate
cortex_m reset_config vectreset
Info : clock speed 2000 kHz
Info : JTAG tap: lpc4337.m4 tap/device found: 0x4ba00477 (mfg: 0x23b (ARM Ltd.), part: 0xba00, ver: 0x4)
Info : JTAG tap: lpc4337.m0 tap/device found: 0x0ba01477 (mfg: 0x23b (ARM Ltd.), part: 0xba01, ver: 0x0)
Info : lpc4337.m4: hardware has 6 breakpoints, 4 watchpoints
Info : lpc4337.m0: hardware has 2 breakpoints, 1 watchpoints
Info : lpc4337.m0: external reset detected
Info : Listening on port 3333 for gdb connections
Info : Listening on port 3334 for gdb connections
dumped 200 bytes in 0.001690s (115.570 KiB/s)

[DEBUG  ] Using read command: "dump_image {{binary}} {address:#x} {length};"
[DEBUG  ] Expanded read command: "dump_image {binary} 0x1a041000 200;"
[DEBUG  ] Running "openocd -c " source [find board/ftdi_lpc4337.cfg]; init; halt; dump_image /tmp/tmp0u62g77h.bin 0x1a041000 200;  exit"".
[DEBUG  ] Open On-Chip Debugger 0.10.0+dev-00689-g6c2020eb (2019-02-10-17:40)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
adapter speed: 2000 kHz
none separate
cortex_m reset_config vectreset
Info : clock speed 2000 kHz
Info : JTAG tap: lpc4337.m4 tap/device found: 0x4ba00477 (mfg: 0x23b (ARM Ltd.), part: 0xba00, ver: 0x4)
Info : JTAG tap: lpc4337.m0 tap/device found: 0x0ba01477 (mfg: 0x23b (ARM Ltd.), part: 0xba01, ver: 0x0)
Info : lpc4337.m4: hardware has 6 breakpoints, 4 watchpoints
Info : lpc4337.m0: hardware has 2 breakpoints, 1 watchpoints
Info : lpc4337.m0: external reset detected
Info : Listening on port 3333 for gdb connections
Info : Listening on port 3334 for gdb connections
dumped 200 bytes in 0.001685s (115.912 KiB/s)

[DEBUG  ] Found 1 apps on the board.
[DEBUG  ]   1. blink
[DEBUG  ] Using program command: "flash write_image erase {{binary}} {address:#x};"
[DEBUG  ] Expanded program command: "flash write_image erase {binary} 0x1a041000;"
[DEBUG  ] Running "openocd -c " source [find board/ftdi_lpc4337.cfg]; init; halt; flash write_image erase /tmp/tmpg79fuqrj.bin 0x1a041000;  exit"".
[DEBUG  ] Open On-Chip Debugger 0.10.0+dev-00689-g6c2020eb (2019-02-10-17:40)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
adapter speed: 2000 kHz
none separate
cortex_m reset_config vectreset
Info : clock speed 2000 kHz
Info : JTAG tap: lpc4337.m4 tap/device found: 0x4ba00477 (mfg: 0x23b (ARM Ltd.), part: 0xba00, ver: 0x4)
Info : JTAG tap: lpc4337.m0 tap/device found: 0x0ba01477 (mfg: 0x23b (ARM Ltd.), part: 0xba01, ver: 0x0)
Info : lpc4337.m4: hardware has 6 breakpoints, 4 watchpoints
Info : lpc4337.m0: hardware has 2 breakpoints, 1 watchpoints
Info : lpc4337.m0: external reset detected
Info : Listening on port 3333 for gdb connections
Info : Listening on port 3334 for gdb connections
auto erase enabled
Warn : Adding extra erase range, 0x00040000 to 0x00040fff
wrote 61440 bytes from file /tmp/tmpg79fuqrj.bin in 1.456898s (41.183 KiB/s)

[DEBUG  ] Erasing page at address 0x1a042000
[DEBUG  ] Using erase command: "flash fillb {address:#x} 0x00 512;"
[DEBUG  ] Expanded erase command: "flash fillb 0x1a042000 0x00 512;"
[DEBUG  ] Running "openocd -c " source [find board/ftdi_lpc4337.cfg]; init; halt; flash fillb 0x1a042000 0x00 512;  exit"".
[DEBUG  ] Open On-Chip Debugger 0.10.0+dev-00689-g6c2020eb (2019-02-10-17:40)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
adapter speed: 2000 kHz
none separate
cortex_m reset_config vectreset
Info : clock speed 2000 kHz
Info : JTAG tap: lpc4337.m4 tap/device found: 0x4ba00477 (mfg: 0x23b (ARM Ltd.), part: 0xba00, ver: 0x4)
Info : JTAG tap: lpc4337.m0 tap/device found: 0x0ba01477 (mfg: 0x23b (ARM Ltd.), part: 0xba01, ver: 0x0)
Info : lpc4337.m4: hardware has 6 breakpoints, 4 watchpoints
Info : lpc4337.m0: hardware has 2 breakpoints, 1 watchpoints
Info : lpc4337.m0: external reset detected
Info : Listening on port 3333 for gdb connections
Info : Listening on port 3334 for gdb connections
wrote 512 bytes to 0x1a042000 in 0.086680s (5.768 KiB/s)

[INFO   ] Finished in 1.689 seconds
```

bradjc commented:
No, nothing obvious.

dc740 commented:
You were right. The "erase" command right after write_image is actually taking down an entire 64kb sector (check the warning in the logs above). I updated the commands, but it's a pretty bad situation, since it's a very big waste of flash. I could not make the write work without the erase.


I'm also seeing that only the first app gets started, I don't see the second one running at all. So, If I compile the examples from the libc repo, and install
1. examples/blink/build/blink.tab
2. `examples/tutorials/02_button_print/build/02_button_print.tab`
I see leds blinking but nothing happens when I press the button (the console works OK, I can see all the kernel debug messages I added)
if I install it in the opposite order... I see output in the uart when I push the button, but the leds are turned off. Does it sound familiar? Did this ever happened when porting a new board?

bradjc commented:
Hmm, what if we had an option where:
1. all apps were read in to tockloader memory off the board
2. the erase happened to erase all app space
3. the old and new apps were written

I think that would be a workaround for this. I'm not sure if it is "better" than enforcing a minimum app size in all cases, but it would be another tool to support certain boards.

dc740 commented:
To be 100% honest... I thought that was what was going on in the background until I saw the openocd logs. And I also thought about implementing it, and then thought you guys wouldn't approve it.

It does have an important drawback because it's going to re-write to the flash all the existing apps... BUT it would allow us to workaround issues like this one, and maybe it could be used only for boards with this kind of bugs to better handle the space. I will still try to find a different workaround anyway.

Another idea would be to allow tockloader to prepare a "collection/package"... which would be a file that contains a set of applications that get installed together. This would be pretty handy once you get something ready and you just want to install a set of apps on a bunch of boards. You would just have to install the packages. But this is just an idea. The workaround you mentioned above is worth investigating

bradjc commented:
I added a `--rewrite-apps` flag that causes an erase to happen before modifying apps in any way. Can you check if that helps in your case? I am a little concerned that the method for erasing flash between reading the existing apps and re-writing everything won't work, and that we will need an additional API in board_interface.py for erasing. But I'm not entirely sure.

Also, if things do work I will update the `BOARDS_APP_DETAILS` feature so that this new board will use `--rewrite-apps` by default without needing the command line argument.

dc740 commented:
OKs, I just tried. It didn't have the intended effects.

Most important::
The 'program' instruction was executed twice on the second app, this means you still call `write_image erase` twice, and as a result, the sector gets cleared when you install the second app. I was hoping there would only be one call, by reading the apps on the board, concatenating it into a single blob together with the new app, and then calling write_image erase only once.

Another side effect of the changes was the heavy use of the 'erase' instructions before the write (`flash fillb {address:#x} 0x00 512;`) which made the whole process relatively slow. I don't fully understand why manually calling fillb and setting the memory to 0 is NOT the same as calling erase in the same flash command with write_image, but in the end you need to use `flash write_image erase` or the program won't get flashed.

dc740 commented:
Check my latest commit. I removed the erase part, which actually just slows down the process and still causes the bug. I made sure there is only 1 call to flash_binary when rewrite-apps is set. We were supposed to flash all apps anyway, so it made no sense to do it on different commands. This way I can go back to the smaller page size as minimum size, and also write several applications one next to the other, saving A LOT of flash space.

side-note: The source code mentions the apps 'linked' list in three places... And that confused me a lot. There is nothing linked about the app list. They are just one after the other in the flash, with no code or reference 'linking' them. Nothing serious, but it can confuse someone else too.

bradjc commented on tockloader/board_interface.py:
```diff
@@ -78,6 +78,16 @@ class BoardInterface:
+		                      'apps_start_address': 0x1a040000,
+		                      'openocd': 'ftdi_lpc4337.cfg',
+		                      'openocd_options': ['noreset'],
+		                      'openocd_commands': {'program': 'flash write_image erase {{binary}} {address:#x};verify_image {{binary}} {address:#x};',
```
I thought it was this "erase" that is the problem. Can't that be removed now that the area is pre-erased?

dc740 replied:
no worries, it thought so too, because it's the logical thing to conclude. Yet the problem is that the write_image won't work without the erase parameter.

I even tried deleting it, and adding a fillb command on this line before the write_image, and even though the flash effectively gets set to 0... the write_image doesn't work without the erase parameter. It is confusing, and probably a bug in openocd, or the flash drivers.

bradjc replied:
Ok gotcha. So it's functionally what you had originally, that the minimum writeable size is a sector. And perhaps openocd could be updated to avoid this, but it is easier to just batch the write using tockloader. I will keep working on this.

bradjc commented:
Ok I finally implemented `--bundle-apps` which should provide the functionality you need.

dc740 commented:
Thanks, I'll pull the changes tonight to test it and update this PR.

bradjc commented:
I put together a branch that should be testable: https://github.com/tock/tockloader/tree/edu-ciaa-nxp

dc740 commented:
Hey! thanks for that! I had a very long day yesterday but I was able to test your branch today and it works perfectly! Do you want me to decline this PR so we can create a new one from your branch?

On a totally different issue:
I'd like to add 'openocd': True to the dict of cmd_flags, but it seems that option doesn't work, since I still have to send it from the command line. Was that intended? (I haven't had the time to read the new implementation yet. probably on the weekend)

bradjc commented:
I updated it so that the channel-related flags can be set by default, and merged that branch.

I'm not totally sure about setting the channel by default, so I didn't add it to other boards. For boards that can only ever use one channel it makes sense. But for others, it might be difficult to know what the default should be, and then any future changes might be breaking changes for users.
